### DISCUSSION TITLE
Is this the correct way to use regex.replace?

### DISCUSSION THREAD

**Discussion by frederikhors** [rust-lang/regex#911] [Q&A] [answered]
I'm trying to write in Rust [this OS answer for javascript](https://stackoverflow.com/a/73973099).

PLAYGROUND here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bb24fc62bc74fa7b3103f238b4660455

```rust
pub fn magic(input: &str) -> String {
    let reg = regex::Regex::new(r"^[ \t]+|[ \t]+$|([ \t]){2,}/mg").unwrap();
    
    reg.replace(input, "$1").to_string()
}

#[cfg(test)]
fn magic_data() -> std::collections::HashMap<&'static str, &'static str> {
    std::collections::HashMap::from([
        ("   ", ""),
        ("    a l    l   lower      ", "a l l lower"),
        (
            "    i need\n new lines\n\n many times     ",
            "i need\nnew lines\n\nmany times",
        ),
        (
            "    i need  \n new lines \n\n many times     ",
            "i need\nnew lines\n\nmany times",
        ),
        (
            "  i need \r\n new lines\r\nmany times   ",
            "i need\nnew lines\nmany times",
        ),
        (
            "    i need \t new lines\t \t many times     ",
            "i need new lines many times",
        ),
        ("  à   la  ", "à la"),
    ])
}

#[test]
fn test() {
    for (k, v) in magic_data() {
        assert_eq!(magic(k), v)
    }
}
```

But as you can see all the tests fail!

**Why?**

**Comment by BurntSushi** [comment, marked as answer]
Here's the first problem that pops out at me:

> `r"^[ \t]+|[ \t]+$|([ \t]){2,}/mg"`

Why are you adding `/mg` to the regex? That only works for *Javascript regex literals*. This is Rust. Rust does not have regex literals. The [docs have several examples for how to set flags](https://docs.rs/regex/latest/regex/#grouping-and-flags). In this case, you'd want `(?m)^[ \t]+|[ \t]+$|([ \t]){2,}`. The `g` flag is irrelevant in Rust, and only exists in Javascript because of some oddities with its regex API.

> `reg.replace(input, "$1").to_string()`

This is your second problem. From the [docs of `Regex::replace`](https://docs.rs/regex/latest/regex/struct.Regex.html#method.replace):

> Replaces the leftmost-first match with the replacement provided.

It only replaces the first match. It looks like you want _all_ matches replaced. So you should switch to [`Regex::replace_all`](https://docs.rs/regex/latest/regex/struct.Regex.html#method.replace_all).

Your third problem is that [`(?m:$)` matches only at the end of the text or right before `\n`, but not immediately before `\r\n`](https://github.com/rust-lang/regex/issues/244). You _also_ appear to have a transcription error from your SO question. Namely, you have this:

```
        (
            "  i need \r\n new lines\r\nmany times   ",
            "i need\nnew lines\nmany times",
        ),
```

But the corresponding SO answer has this:

```
        (
            "  i need \r\n new lines\r\nmany times   ",
            "i need\r\nnew lines\r\nmany times",
        ),
```

Fixing the CRLF issue in the regex is a little tricky, but it can be done:

```
    let reg = regex::Regex::new(r"(?m)^[ \t]+|[ \t]+(\r?)$|([ \t]){2,}").unwrap();
    reg.replace_all(input, "$1$2").to_string()
```

Given all of that, your tests pass: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8e63bdc73c1c0fa2f6412c78a119f962

**Reply by frederikhors** [reply]
Thank you for everything! Is the first time with Rust for me!

The link you posted doesn't handle the case:

```rust
(
  "  i need \r\n new lines\r\nmany times   ",
  "i need\nnew lines\nmany times",
)
```

In your playground this is written like this:

```rust
(
  "  i need \r\n new lines\r\nmany times   ",
  "i need\r\nnew lines\r\nmany times",
),
```

How to fix this (replace `\r\n` to `\n`)?

**Reply by BurntSushi** [reply]
Like I said in the comment, that didn't match the test in the SO post. But if you just want to normalize `\r\n` to `\n`, then, well, do that? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4364fa729fe708afc83b1565490982e9
